File: src/main/java/mods/cybercat/gigeresque/common/config/ConfigAccessor.java
package mods.cybercat.gigeresque.common.config;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import mods.cybercat.gigeresque.common.entity.EntityIdentifiers;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;

public class ConfigAccessor {
	private static Map<String, String> reversedMorphMappings;
	private static Map<ResourceLocation, List<String>> whitelistMappings;
	private static Map<ResourceLocation, List<String>> blacklistMappings;

	private synchronized static Map<ResourceLocation, List<String>> getWhitelistMappings() {
		if (whitelistMappings == null)
			whitelistMappings = Map.of(EntityIdentifiers.ALIEN, GigeresqueConfig.alienWhitelist,
					EntityIdentifiers.AQUATIC_ALIEN, GigeresqueConfig.aquaticAlienWhitelist,
					EntityIdentifiers.FACEHUGGER, GigeresqueConfig.facehuggerWhitelist, EntityIdentifiers.RUNNER_ALIEN,
					GigeresqueConfig.runnerWhitelist);
		return whitelistMappings;
	}

	private synchronized static Map<ResourceLocation, List<String>> getBlacklistMappings() {
		if (blacklistMappings == null)
			blacklistMappings = Map.of(EntityIdentifiers.ALIEN, GigeresqueConfig.alienBlacklist,
					EntityIdentifiers.AQUATIC_ALIEN, GigeresqueConfig.aquaticAlienBlacklist,
					EntityIdentifiers.FACEHUGGER, GigeresqueConfig.facehuggerBlacklist, EntityIdentifiers.RUNNER_ALIEN,
					GigeresqueConfig.runnerBlacklist,EntityIdentifiers.MUTANT_POPPER,
					GigeresqueConfig.smallMutantHosts,EntityIdentifiers.MUTANT_HAMMERPEDE,
					GigeresqueConfig.smallMutantHosts,EntityIdentifiers.MUTANT_STALKER,
					GigeresqueConfig.largeMutantHosts);
		return blacklistMappings;
	}

	public static boolean isTargetWhitelisted(LivingEntity entity, Entity target) {
		return isTargetWhitelisted(entity.getClass(), target);
	}

	public static boolean isTargetWhitelisted(Class<? extends Entity> entityClass, Entity target) {
		if (target == null)
			return false;
		var attackerIdentifier = EntityIdentifiers.typeMappings.get(entityClass);
		var whitelist = getWhitelistMappings().getOrDefault(attackerIdentifier, Collections.emptyList());
		var targetIdentifier = BuiltInRegistries.ENTITY_TYPE.getKey(target.getType());
		return whitelist.contains(targetIdentifier.toString());
	}

	public static boolean isTargetBlacklisted(LivingEntity entity, Entity target) {
		return isTargetBlacklisted(entity.getClass(), target);
	}

	public static boolean isTargetHostable(Entity target) {
		var normalEntries = GigeresqueConfig.alienHosts;
		var waterEntries = GigeresqueConfig.aquaticAlienHosts;
		var runnerEntries = GigeresqueConfig.runnerHosts;
		var targetIdentifier = BuiltInRegistries.ENTITY_TYPE.getKey(target.getType());
		return normalEntries.contains(targetIdentifier.toString()) || waterEntries.contains(targetIdentifier.toString())
				|| runnerEntries.contains(targetIdentifier.toString());
	}

	public static boolean isTargetSmallMutantHost(Entity target) {
		var waveEntries = GigeresqueConfig.smallMutantHosts;
		var targetIdentifier = BuiltInRegistries.ENTITY_TYPE.getKey(target.getType());
		return waveEntries.contains(targetIdentifier.toString());
	}

	public static boolean isTargetLargeMutantHost(Entity target) {
		var waveEntries = GigeresqueConfig.largeMutantHosts;
		var targetIdentifier = BuiltInRegistries.ENTITY_TYPE.getKey(target.getType());
		return waveEntries.contains(targetIdentifier.toString());
	}

	public static boolean isTargetAlienHost(Entity target) {
		var waveEntries = GigeresqueConfig.alienHosts;
		var targetIdentifier = BuiltInRegistries.ENTITY_TYPE.getKey(target.getType());
		return waveEntries.contains(targetIdentifier.toString());
	}

	public static boolean isTargetDNAImmune(Entity target) {
		var waveEntries = GigeresqueConfig.dnaBlacklist;
		var targetIdentifier = BuiltInRegistries.ENTITY_TYPE.getKey(target.getType());
		return waveEntries.contains(targetIdentifier.toString());
	}

	public static boolean isTargetBlacklisted(Class<? extends Entity> entityClass, Entity target) {
		if (target == null)
			return false;
		var attackerIdentifier = EntityIdentifiers.typeMappings.get(entityClass);
		var blacklist = getBlacklistMappings().getOrDefault(attackerIdentifier, Collections.emptyList());
		var targetIdentifier = BuiltInRegistries.ENTITY_TYPE.getKey(target.getType());
		return blacklist.contains(targetIdentifier.toString());
	}

	public static synchronized Map<String, String> getReversedMorphMappings() {
		if (reversedMorphMappings == null)
			processReversedMorphMappings();
		return reversedMorphMappings;
	}

	private static void processReversedMorphMappings() {
		HashMap<String, String> map = new HashMap<>();

		var internalMap = Map.of(EntityIdentifiers.ALIEN.toString(), GigeresqueConfig.alienHosts,
				EntityIdentifiers.AQUATIC_ALIEN.toString(), GigeresqueConfig.aquaticAlienHosts,
				EntityIdentifiers.RUNNER_ALIEN.toString(), GigeresqueConfig.runnerHosts);

		internalMap.forEach((morphTo, morphFromSet) -> morphFromSet
				.forEach(morphFrom -> map.computeIfAbsent(morphFrom, (it) -> morphTo)));
		reversedMorphMappings = map;
	}

}

File: src/main/kotlin/com/bvanseg/gigeresque/common/config/ConfigAccessor.kt
package com.bvanseg.gigeresque.common.config

import com.bvanseg.gigeresque.common.Gigeresque
import com.bvanseg.gigeresque.common.entity.EntityIdentifiers
import net.minecraft.entity.Entity
import net.minecraft.entity.LivingEntity
import net.minecraft.util.registry.Registry
import java.util.*
import kotlin.reflect.KClass

/**
 * @author Boston Vanseghi
 */
object ConfigAccessor {

    val mappedAcidResistantBlocks: Map<String, Set<String>> by lazy { processAcidResistantBlocks() }

    val reversedMorphMappings: Map<String, String> by lazy { processReversedMorphMappings() }

    private val whitelistMappings by lazy {
        mapOf(
            EntityIdentifiers.ALIEN to Gigeresque.config.targeting.alienWhitelist,
            EntityIdentifiers.AQUATIC_ALIEN to Gigeresque.config.targeting.aquaticAlienWhitelist,
            EntityIdentifiers.FACEHUGGER to Gigeresque.config.targeting.facehuggerWhitelist,
            EntityIdentifiers.RUNNER_ALIEN to Gigeresque.config.targeting.runnerWhitelist
        )
    }

    private val blacklistMappings by lazy {
        mapOf(
            EntityIdentifiers.ALIEN to Gigeresque.config.targeting.alienBlacklist,
            EntityIdentifiers.AQUATIC_ALIEN to Gigeresque.config.targeting.aquaticAlienBlacklist,
            EntityIdentifiers.FACEHUGGER to Gigeresque.config.targeting.facehuggerBlacklist,
            EntityIdentifiers.RUNNER_ALIEN to Gigeresque.config.targeting.runnerBlacklist
        )
    }

    fun isTargetWhitelisted(entity: LivingEntity, target: Entity?) = isTargetWhitelisted(entity::class, target)
    fun isTargetBlacklisted(entity: LivingEntity, target: Entity?) = isTargetBlacklisted(entity::class, target)

    fun isTargetWhitelisted(entityClass: KClass<*>, target: Entity?): Boolean {
        if (target == null) return false
        val attackerIdentifier = EntityIdentifiers.typeMappings[entityClass] ?: return false
        val whitelist = whitelistMappings[attackerIdentifier] ?: return false
        val targetIdentifier = Registry.ENTITY_TYPE.getId(target.type)
        return whitelist.contains(targetIdentifier.toString())
    }

    fun isTargetBlacklisted(entityClass: KClass<*>, target: Entity?): Boolean {
        if (target == null) return false
        val attackerIdentifier = EntityIdentifiers.typeMappings[entityClass] ?: return false
        val blacklist = blacklistMappings[attackerIdentifier] ?: return false
        val targetIdentifier = Registry.ENTITY_TYPE.getId(target.type)
        return blacklist.contains(targetIdentifier.toString())
    }

    private fun processAcidResistantBlocks(): Map<String, Set<String>> {
        val map = hashMapOf<String, HashSet<String>>()
        Gigeresque.config.miscellaneous.acidResistantBlocks.forEach {
            val parts = it.lowercase(Locale.US).split(":")

            if (parts.size == 1) {
                map.computeIfAbsent("minecraft") { hashSetOf() }.add(parts[0])
            } else if (parts.size == 2) {
                map.computeIfAbsent(parts[0]) { hashSetOf() }.add(parts[1])
            }
        }

        return map
    }

    private fun processReversedMorphMappings(): Map<String, String> {
        val map = hashMapOf<String, String>()

        val internalMap = mapOf(
            EntityIdentifiers.ALIEN.toString() to Gigeresque.config.morphing.alienHosts,
            EntityIdentifiers.AQUATIC_ALIEN.toString() to Gigeresque.config.morphing.aquaticAlienHosts,
            EntityIdentifiers.RUNNER_ALIEN.toString() to Gigeresque.config.morphing.runnerHosts
        )

        internalMap.forEach { (morphTo, morphFromSet) ->
            morphFromSet.forEach { morphFrom ->
                map.computeIfAbsent(morphFrom) { morphTo }
            }
        }
        return map
    }
}